odoo/models.py

```diff
@@ -5657,28 +5657,35 @@ def _cache(self):
         """ Return the cache of ``self``, mapping field names to values. """
         return RecordCache(self)
 
-    @api.model
     def _in_cache_without(self, field, limit=PREFETCH_MAX):
         """ Return records to prefetch that have no value in cache for ``field``
             (:class:`Field` instance), including ``self``.
             Return at most ``limit`` records.
         """
+        recs = self.browse(unique(self._prefetch_ids))
+        ids = self.env.cache.get_missing_ids(recs, field)
         # This method returns records that are either all real, or all new.
         # Those records are aimed at being either fetched, or computed.  But the
         # method '_fetch_field' is not correct with new records: it considers
         # them as forbidden records, and clears their cache!  On the other hand,
         # compute methods are not invoked with a mix of real and new records for
         # the sake of code simplicity.
+        return self._batch_with(ids, limit=limit)
+
+    def _batch_with(self, ids, limit=PREFETCH_MAX):
+        """ Return a batch of at most ``limit`` records from the record ``self``
+        and a subset of ``ids``.  The batch is such that its records are either
+        all real, or all new.
+        """
```

Comment:
Does this actually belong in BaseModel? It does rely on some ORM guarantees (e.g. real vs new ids) but doesn't seem to have any reason too be a BaseModel method.

The naming and API also don't really hint at the semantics.

Maybe the method should be split into two utilities to combine e.g. `expand_from(origin, expand_source, limit)` which would expand the `origin` set to `limit` using the expansion source, and `filter_kind` which would... filter by kind? (lazily) Not sure the latter is really needed but it could be useful to avoid the risk of repeatedly evaluated the reference kind.